### ISSUE TITLE
cupy.sum() only using 1 block with 512 threads?

### ISSUE THREAD

**Issue by NaderAlAwar** [cupy/cupy#6512]
### Description

Calling cupy.sum() on an array with any length will only use 1 block with 512 threads according to nvprof, even for larger lengths. Shouldn't the number of threads scale with the size of the array to improve performance?

### To Reproduce

```py
# Write the code here
cp_array = cp.random.rand(100000) # can be 100, 250000, etc.

for i in range(100):
    cp.sum(cp_array)
```

Running `nvprof --print-gpu-trace python cupy_test.py` will show:

```
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream  Name

1.15826s  4.8320us              (1 1 1)       (512 1 1)        22  4.0000KB        0B         -           -           -           -  Tesla V100-SXM2         1         7  cupy_sum [1291]

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
# Paste the output here
OS                           : Linux-4.14.0-115.21.2.1chaos.ch6a.ppc64le-ppc64le-with-glibc2.17
Python Version               : 3.9.7
CuPy Version                 : 10.1.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.20.3
SciPy Version                : 1.7.3
Cython Build Version         : 0.29.27
Cython Runtime Version       : None
CUDA Root                    : /usr/workspace/alawar1/projects/anaconda3/envs/boltz_pool
nvcc PATH                    : None
CUDA Build Version           : 10020
CUDA Driver Version          : 10020
CUDA Runtime Version         : 10020
cuBLAS Version               : (available)
cuFFT Version                : 10102
cuRAND Version               : 10102
cuSOLVER Version             : (10, 3, 0)
cuSPARSE Version             : (available)
NVRTC Version                : (10, 2)
Thrust Version               : 100907
CUB Build Version            : <unknown>
Jitify Build Version         : 6dbba68
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla V100-SXM2-16GB
Device 0 Compute Capability  : 70
Device 0 PCI Bus ID          : 0004:04:00.0
Device 1 Name                : Tesla V100-SXM2-16GB
Device 1 Compute Capability  : 70
Device 1 PCI Bus ID          : 0004:05:00.0
Device 2 Name                : Tesla V100-SXM2-16GB
Device 2 Compute Capability  : 70
Device 2 PCI Bus ID          : 0035:03:00.0
Device 3 Name                : Tesla V100-SXM2-16GB
Device 3 Compute Capability  : 70
Device 3 PCI Bus ID          : 0035:04:00.0
```


### Additional Information

_No response_

**Comment by leofang** [comment]
No it's not how things work for GPU. There's a hardware limitation in the block size (number of threads per block, which is 2048 on NVIDIA GPU IIRC) and making it too large usually is not helping (reduces the occupancy). Reduction kernels like `sum()` usually requires two or more passes to complete in terms of both algorithm design & implementation, so certainly it's not making sense to indefinitely scale up the block size.

However, there're some improvement plans being discussed, ex: https://github.com/cupy/cupy/issues/5194. But for now it's easiest for you to try the CUB or cuTENSOR backend: https://docs.cupy.dev/en/stable/user_guide/performance.html#use-cub-cutensor-backends-for-reduction-and-other-routines

**Comment by NaderAlAwar** [comment]
Using the CUB backend works. Thanks! You can close this

**Comment by NaderAlAwar** [comment]
Actually, I have a follow up question. I understand that the block size is limited, but couldn't you just increase the number of blocks?

**Comment by leofang** [comment]
Yeah I realized the grid size might be what you were actually asking 😅 I don't remember the detail off top of my head (and am not in front of my computer), maybe a team member could chime in later.
